tests/interfaces/test_autograd_default_qubit_2.py

```diff
@@ -0,0 +1,689 @@
+        assert np.allclose(res[3], np.cos(x) * np.cos(y))
+
+        jac = qml.jacobian(cost)(params)
+        assert isinstance(jac, np.ndarray)
+        assert jac.shape == (4, 2)  # pylint: disable=no-member
+
+        assert np.allclose(jac[1:3], 0)
+
+        d1 = -np.sin(x) * np.cos(y)
+        assert np.allclose(jac[0, 0], d1)
+        assert np.allclose(jac[3, 0], d1)
+
+        d2 = -np.cos(x) * np.sin(y)
+        assert np.allclose(jac[0, 1], d2)
+        assert np.allclose(jac[3, 1], d2)
+
+    def test_reusing_quantum_tape(self, execute_kwargs, tol):
+        """Test re-using a quantum tape by passing new parameters"""
+        a = np.array(0.1, requires_grad=True)
+        b = np.array(0.2, requires_grad=True)
+
+        dev = DefaultQubit2()
+
+        tape = qml.tape.QuantumScript(
+            [qml.RY(a, 0), qml.RX(b, 1), qml.CNOT((0, 1))],
+            [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(1))],
+        )
+        assert tape.trainable_params == [0, 1]
+
+        def cost(a, b):
+            tape.set_parameters([a, b])
+            return autograd.numpy.hstack(execute([tape], dev, **execute_kwargs)[0])
+
+        jac_fn = qml.jacobian(cost)
+        jac = jac_fn(a, b)
+
+        a = np.array(0.54, requires_grad=True)
+        b = np.array(0.8, requires_grad=True)
+
+        # check that the cost function continues to depend on the
+        # values of the parameters for subsequent calls
+        res2 = cost(2 * a, b)
+        expected = [np.cos(2 * a), -np.cos(2 * a) * np.sin(b)]
+        assert np.allclose(res2, expected, atol=tol, rtol=0)
+
+        jac_fn = qml.jacobian(lambda a, b: cost(2 * a, b))
+        jac = jac_fn(a, b)
+        expected = (
+            [-2 * np.sin(2 * a), 2 * np.sin(2 * a) * np.sin(b)],
+            [0, -np.cos(2 * a) * np.cos(b)],
+        )
+        assert isinstance(jac, tuple) and len(jac) == 2
+        assert all(np.allclose(_j, _e, atol=tol, rtol=0) for _j, _e in zip(jac, expected))
+
+    def test_classical_processing(self, execute_kwargs):
+        """Test classical processing within the quantum tape"""
+        a = np.array(0.1, requires_grad=True)
+        b = np.array(0.2, requires_grad=False)
+        c = np.array(0.3, requires_grad=True)
+
+        def cost(a, b, c, device):
+            ops = [
+                qml.RY(a * c, wires=0),
+                qml.RZ(b, wires=0),
+                qml.RX(c + c**2 + np.sin(a), wires=0),
+            ]
+
+            tape = qml.tape.QuantumScript(ops, [qml.expval(qml.PauliZ(0))])
+            return execute([tape], device, **execute_kwargs)[0]
+
+        dev = DefaultQubit2()
+        res = qml.jacobian(cost)(a, b, c, device=dev)
+
+        # Only two arguments are trainable
+        assert isinstance(res, tuple) and len(res) == 2
+        assert res[0].shape == ()
+        assert res[1].shape == ()
+
+        # I tried getting analytic results for this circuit but I kept being wrong and am giving up
```

Comment:
appreciate the notice and I'm not bothered by this, but I'd probably remove it before merging :D